components/CryptoSection.tsx

```diff
@@ -0,0 +1,139 @@
+
+import { Button } from "@components/ui/Button";
+
+import { useContracts } from "../contexts/contractsContext";
+import genericAbi from "../web3/abis/abiWithGetBalance.json";
+
+interface Window {
+  ethereum: any;
+  web3: Web3;
+}
+
+interface EvtsToVerify {
+  [eventId: string]: boolean;
+}
+
+declare const window: Window;
+
+interface CryptoSectionProps {
+  id: number | string;
+  smartContractAddress: string;
+  /** When set to true, there will be only 1 button which will both connect Metamask and verify the user's wallet. Otherwise, it will be in 2 steps with 2 buttons. */
+  oneStep: boolean;
+  verified: boolean | undefined;
+  setEvtsToVerify: React.Dispatch<React.SetStateAction<Record<number | string, boolean>>>;
+}
+
+const CryptoSection = (props: CryptoSectionProps) => {
+  // Crypto section which should be shown on booking page if event type requires a smart contract token.
+  const [ethEnabled, toggleEthEnabled] = useState<boolean>(false);
+  const { addContract } = useContracts();
+
+  const connectMetamask = useCallback(async () => {
+    if (window.ethereum) {
+      await window.ethereum.request({ method: "eth_requestAccounts" });
+      window.web3 = new Web3(window.ethereum);
+      toggleEthEnabled(true);
+    } else {
+      toggleEthEnabled(false);
+    }
+  }, []);
+
+  const verifyWallet = useCallback(async () => {
+    try {
+      const contract = new window.web3.eth.Contract(genericAbi as AbiItem[], props.smartContractAddress);
+      const balance = await contract.methods.balanceOf(window.ethereum.selectedAddress).call();
+
+      const hasToken = balance > 0;
+
+      if (!hasToken) {
+        throw new Error("Specified wallet does not own any tokens belonging to this smart contract");
+      } else {
+        const account = (await web3.eth.getAccounts())[0];
+
+        const signature = await web3.eth.personal.sign(process.env.NEXT_PUBLIC_WEB3_AUTH_MSG, account);
+        addContract({ address: props.smartContractAddress, signature });
+
+        await verifyAccount(signature, account);
+
+        props.setEvtsToVerify((prevState: EvtsToVerify) => {
+          const changedEvt = { [props.id]: hasToken };
+          return { ...prevState, ...changedEvt };
+        });
+      }
+    } catch (err) {
+      err instanceof Error ? showToast(err.message, "error") : showToast("An error has occurred", "error");
+    }
+  }, [props, addContract]);
+
+  // @TODO: Show error on either of buttons if fails. Yup schema already contains the error message.
+  const successButton = useMemo(() => {
+    return (
+      <Button type="button" disabled>
+        Success!
+      </Button>
+    );
+  }, []);
+
+  const verifyButton = useMemo(() => {
+    return (
+      <Button color="secondary" onClick={verifyWallet} type="button" id="hasToken" name="hasToken">
+        <img className="h-5 mr-1" src="/integrations/metamask.svg" /> Verify wallet
+      </Button>
+    );
+  }, [verifyWallet]);
```

Comment:
Why not just write idiomatic React and create a component for this?
Or just inline where it's used as it's so small.

Feels like a lot more code in this component than is needed.